motif/motif.go

```diff
@@ -0,0 +1,181 @@
+
+import (
+	"fmt"
+	"github.com/vertgenlab/gonomics/common"
+	"github.com/vertgenlab/gonomics/exception"
+	"github.com/vertgenlab/gonomics/fileio"
+	"log"
+	"strings"
+)
+
+type PositionMatrixType byte
+
+const (
+	Frequency   PositionMatrixType = 0
+	Probability PositionMatrixType = 1
+	Weight      PositionMatrixType = 2
+	None        PositionMatrixType = 3
+)
+
+//PositionMatrix is a struct encoding a position frequency/probability/weight matrix.
+type PositionMatrix struct {
+	Id   string
+	Name string
+	Type PositionMatrixType
+	Mat  [][]float64
+}
+
+// Write writes a slice of PositionMatrix structs to an output filename.
+func Write(filename string, records []PositionMatrix) {
+	var err error
+	file := fileio.EasyCreate(filename)
+	for _, m := range records {
+		WritePositionMatrix(file, m)
+	}
+	err = file.Close()
+	exception.PanicOnErr(err)
+}
+
+// WritePositionMatrix writes an individual PositionMatrix struct to a fileio.EasyWriter.
+func WritePositionMatrix(file *fileio.EasyWriter, m PositionMatrix) {
+	var err error
+	_, err = fmt.Fprintf(file, ">%s\t%s\n%s", m.Id, m.Name, matToString(m.Mat))
+	exception.PanicOnErr(err)
+}
+
+//matToString is a helper function of matToString that converts a Pfm/Ppm matrix to a string.
+func matToString(mat [][]float64) string {
+	var answer string = "A\t[\t"
+	if len(mat) != 4 {
+		log.Fatalf("Error: Input PFM must have 4 rows, one for each nucleotide.")
+	}
+	for i := range mat[0] {
+		answer = answer + fmt.Sprintf("\t%g", mat[0][i])
+	}
+	answer = answer + "\t]\nC [ "
+	for i := range mat[1] {
+		answer = answer + fmt.Sprintf("\t%g", mat[1][i])
+	}
+	answer = answer + "\t]\nG [ "
+	for i := range mat[2] {
+		answer = answer + fmt.Sprintf("\t%g", mat[2][i])
+	}
+	answer = answer + "\t]\nT [ "
+	for i := range mat[3] {
+		answer = answer + fmt.Sprintf("\t%g", mat[3][i])
+	}
+	answer = answer + "\t]\n"
+	return answer
+}
+
+//StringToPositionMatrixType parses a jaspar.PositionMatrixType type from an input string.
+func StringToPositionMatrixType(s string) PositionMatrixType {
+	switch s {
+	case "Frequency":
+		return Frequency
+	case "Probability":
+		return Probability
+	case "Weight":
+		return Weight
+	default:
+		log.Fatalf("Error: expected %s to be a position matrix type that is either 'Frequency', 'Probability', or 'Weight'.\n", s)
+		return None
+	}
+}
+
+//Read parses a slice of PositionMatrix structs from an input file in JASPAR format.
+func Read(filename string, Type string) []PositionMatrix {
+	var err error
+	var curr PositionMatrix
+	var answer []PositionMatrix
+	var doneReading bool
+	usedMotifIds := make(map[string]bool)
+
+	file := fileio.EasyOpen(filename)
+
+	for curr, doneReading = NextPositionMatrix(file, StringToPositionMatrixType(Type)); !doneReading; curr, doneReading = NextPositionMatrix(file, StringToPositionMatrixType(Type)) {
+		if usedMotifIds[curr.Id] {
+			log.Fatalf("Error: %s is used as the ID for multiple records. IDs must be unique.", curr.Id)
+		} else {
```

Comment:
not a bug, but the "else" is not needed since you have a fatal in the "if"

Reply:
good catch!